platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,669 @@
+of transmission by signalling to the sender how many bytes can be sent before receiving a new update from the receiver.
+The same technique is used for backpressure in the HTTP2 protocol. We will use the same concept here.
+
+Execution will instruct Consensus each time it needs additional rounds processed. It could indicate any number of
+additional rounds. For each round, it will produce the appropriate roster, even if the roster doesn't change between
+rounds. (And if we have any other dynamic configuration, it is also provided in like manner). Execution is therefore
+responsible for dictating the rate at which rounds can be produced, providing natural backpressure. In addition, by
+sending the roster information for each round, a quick and efficient mechanism is provided for deterministically
+changing the roster for any given round. This is very useful for punishing malicious nodes.
+
+By keeping Consensus tied to Execution in this way, if one node's Execution runs slowly compared to other nodes in the
+network, that node will naturally handle consensus slower than the others, and will therefore eventually fall behind
+and need to reconnect. Indeed, from the perspective of the rest of the consensus module, or from the other nodes in
+the network, the behavior of the node in stress is **exactly the same** whether Consensus or Execution are the reason
+for falling behind.
+
+##### A Silly Example
+
+Suppose we have a 4 node network, where 1 node is a Raspberry PI and the other 3 nodes are 64-core 256GB machines. In
+this network, at 10 TPS, all 4 nodes may be able to process events and handle transactions without any problems. If the
+transaction load were to increase to 10,000 TPS, then the Raspberry PI may not be able to keep up with this workload,
+while the other 3 machines might. The healthy machines will continue to accept transactions and create new events, while
+the machine under stress is unable to create new consensus rounds fast enough, since consensus is stalled waiting for
+the Execution layer to finish processing previous rounds. As time goes on, the slow machine begins to receive events
+from the other nodes which are too far in the future. It drops these events. When it finally needs those events, gossip
+will fetch them again.
+
+It may be that the load decreases back to a manageable level, and the Raspberry PI is able to fetch the events it wasn't
+able to process quickly enough, and successfully "catch up" with the other nodes. Or, it may be that the load continues
+long enough that the Raspberry PI falls so far behind that when it asks the other nodes for events it is missing, that
+those nodes will reply that those events had expired, and the Raspberry PI has truly fallen behind. At this point, the
+only recourse is for the node to reconnect. This it will do, using an exponential backoff algorithm, so that if the PI
```

Comment:
True, if we do reconnect from node to node. Since reconnect is not defined in this document, I didn't go into it, but in my mind reconnect will use block nodes, and nodes will have to pay for that (or use their own block node so it is free) and we then don't have to deal with a malicious reconnect attempt. Block nodes could have such logic, or they could just be happy to help anyone reconnect who is happy to pay (with their own global rate limits in place). But this is a detail we can get into when doing reconnect design on Execution.